File: typescript/tco.ts
#!/usr/bin/env node

// Check Tail Call Optimization support

function fib_tail_recur(n: number): number {
  function visit(n: number, a: number, b: number) {
    if (n == 0) {
      return a;
    } else {
      return visit(n - 1, a + b, a);
    }
  }
  return visit(n, 1, 0);
}

function fib_recur(n: number): number {
  if (n <= 1) {
    return 1;
  } else {
    return fib_recur(n - 1) + fib_recur(n - 2);
  }
}

function fib_iterate(n: number): number {
  let data = [1, 1];

  for (let i = 1; i < n; i++) {
    data = [data[0] + data[1], data[0]];
  }
  return data[0];
}

let n = 0;
console.log(`fib_tail_recur(${n}) ${fib_tail_recur(n)}`);
n = 1;
console.log(`fib_tail_recur(${n}) ${fib_tail_recur(n)}`);
n = 11;
console.log(`fib_tail_recur(${n}) ${fib_tail_recur(n)}`);
n = 1111;
console.log(`fib_tail_recur(${n}) ${fib_tail_recur(n)}`);

n = 0;
console.log(`fib_iterate(${n}) ${fib_iterate(n)}`);
n = 1;
console.log(`fib_iterate(${n}) ${fib_iterate(n)}`);
n = 11;
console.log(`fib_iterate(${n}) ${fib_iterate(n)}`);
n = 1111;
console.log(`fib_iterate(${n}) ${fib_iterate(n)}`);

n = 0;
console.log(`fib_recur(${n}) ${fib_recur(n)}`);
n = 1;
console.log(`fib_recur(${n}) ${fib_recur(n)}`);
n = 11;
console.log(`fib_recur(${n}) ${fib_recur(n)}`);
n = 1111;
console.log(`fib_recur(${n}) ${fib_recur(n)}`);

File: python/tco.py
#!/usr/bin/env python

# check Tail Call Optimization support


def fib_tail_recur(n: int) -> int:
    """
    fib (tail recur version)

    this would check if TCO
    """

    def visit(i: int, a: int, b: int):
        if i == 0:
            return a
        return visit(i - 1, a + b, a)

    return visit(n, 1, 0)


def fib_recur(n: int) -> int:
    """
    fib (recur version)
    """
    if n <= 1:
        return 1
    return fib_recur(n - 1) + fib_recur(n - 2)


def fib(n: int) -> int:
    """
    fib (iteration version)
    """
    a, b = 1, 1
    for _ in range(1, n):
        a, b = a + b, a

    return a


n = 1111
print(f"fib({n})={fib(n)}")
print(f"fib_tail_recur({n})={fib_tail_recur(n)}")
print(f"fib_recur({n})={fib_recur(n)}")
